File: LeetCode/LeetCode-CN-CPP/2020六月每日一题/739. 每日温度.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        vector<int> result(T.size(),0);
        stack<int> stk;

        for(int i = T.size() - 1; i >= 0; i--) {
            // 栈中只需要保存比当前气温高的位置
            while(!stk.empty() && T[i] >= T[stk.top()]) {
                stk.pop(); // 出栈
            }
            // 如果栈中为空，则说明没有比当前气温高，否则计算出最近比当前大的位置与当前的位置
            result[i] = stk.empty() ? 0 : stk.top() - i;
            stk.push(i);
        }

        return result;
    }
};

File: LeetCode/LeetCode-CN-JAVA/2020六月每日一题/739. 每日温度.java
class Solution {
    public int[] dailyTemperatures(int[] T) {
        int[] ans = new int[T.length];
        Stack<Integer> stk = new Stack<Integer>();
        for(int i = T.length - 1; i >= 0; i--) {
            // 栈中保存比当前气温高的位置
            while(!stk.empty() && T[i] >= T[stk.peek()]) {
                stk.pop();
            }
            // 如果栈为空则说明没有比当前气温高，记为0.否则计算出最近一个比当前气温高距离目前的距离。
            ans[i] = stk.empty() ? 0 : stk.peek() - i;
            stk.push(i);
        }
        return ans;
    }
}